### DISCUSSION TITLE
Setting up goma server

### DISCUSSION THREAD

**Discussion by uazo** [bromite/bromite#1032] [Show and tell]
I begin by saying that everything I write here is the result of tests and analysis of the source codes, nothing is to be considered 100% reliable and obviously official.
There are things that I have deliberately not considered, because as a first step I simply wanted to evaluate the possibility of use.
The things deliberately left out are security, understood as SSL and OAUTH, and monitoring (or what I thought is referred to as the statistics module), which they should be dealt with if the thing is to be used on a server exposed to the internet.
There may also be other things, which to date, I have not yet understood.

What I understand:
- goma is a client / server is a system based on gRpc and Google RBE (remote build execution)
  the client (in c++) is quite old (and well done, as always for c++ projects). the server, in the version written in go, is published as if it were a first version, that although it works well, but it may still be considered unstable
- goma opensource server needs an rbe server to work.
   i did not find anything about the official version used by google but i think this is https://cloud.google.com/build/docs/cloud-builders (his address is remotebuildexecution.googleapis.com) and to use it it is necessary to request authorization from google through a particular procedure: we don't care, because in our case we will use `buildgrid`, considering it implements Google's REAPI and RWAPI (which although I don't know what they are, but it seems like the right choice, although I find it very unstable)
- the server component is `remoteexec_proxy` https://chromium.googlesource.com/infra/goma/server/+/refs/heads/master/cmd/remoteexec_proxy/ and includes everything that is necessary for a correct setup: from what I have deduced, it is composed of some modules
    - auth_server
    - cache_server
    - exec_server
    - execlog_server
    - file_server
    - frontend
 which seems to be possible to launch individually
- the client component is `compiler_proxy`, which is started as a daemon by the `goma_ctl.py` script
   it is highly configurable and all flags are available here https://chromium.googlesource.com/infra/goma/client/+/refs/heads/master/client/goma_flags.cc
    the configuration is done through environment variables, by putting "GOMA_" before the flags

let's begin.

# BuildGrid installation https://gitlab.com/BuildGrid/buildgrid
## Prerequisites:
`sudo apt install python3-venv`
## Installation:
```
cd ~
git clone https://gitlab.com/BuildGrid/buildgrid.git
cd buildgrid
python3 -m venv env
env/bin/python -m pip install --upgrade setuptools pip wheel
env/bin/python -m pip install --editable .
```
Change the settings in "data/config/default.yml":
```
server:
  - !channel
    address: localhost:50051
    insecure-mode: true

description: >
  BuildGrid's default configuration:
    - Unauthenticated plain HTTP at :50051
    - Single instance: [unnamed]
    - In-memory data, max. 2Gio
    - DataStore: sqlite:///./example.db
    - Hosted services:
       - ActionCache
       - Execute
       - ContentAddressableStorage
       - ByteStream

authorization:
  method: none

monitoring:
  enabled: false

instances:
  - name: 'default_instance'
    description: |
      The unique '' instance.

    storages:
      - !disk-storage &cas-storage
        path: !expand-path $HOME/cas

    schedulers:
      - !memory-scheduler &state-database
        storage: *cas-storage

    caches:
      - !lru-action-cache &build-cache
        storage: *cas-storage
        max-cached-refs: 256
        cache-failed-actions: true
        allow-updates: true

    services:
      - !action-cache
        cache: *build-cache

      - !execution
        storage: *cas-storage
        action-cache: *build-cache
        scheduler: *state-database
        max-execution-timeout: 7200
        property-keys:
          - dockerPrivileged
          - InputRootAbsolutePath

      - !cas
        storage: *cas-storage

      - !bytestream
        storage: *cas-storage
```
you can try running the server with `env/bin/bgd server start data/config/default.yml -vvv`

you have to start execution worker with:
```
env/bin/bgd bot --instance-name=default_instance --remote=http://localhost:50051 -w OSFamily Linux -vvv host-tools
```
make sure you have all the prerequisites by running this https://chromium.googlesource.com/chromium/src/+/master/build/install-build-deps.sh?format=TEXT

# goma server installation
## prerequisite number 1: go >=1.16.2
```
cd ~
wget https://golang.org/dl/go1.16.2.linux-amd64.tar.gz
sudo rm -rf /usr/local/go
sudo tar -C /usr/local -xzf go1.16.2.linux-amd64.tar.gz
```
try with: `/usr/local/go/bin/go version`
## prerequisite number 2: redis
```
sudo apt install redis-server
```
change with `sudo nano /etc/redis/redis.conf` flag `supervised` from `no` to `systemd`
start with:
`sudo systemctl restart redis.service`
## download sources:
```
cd ~
git clone https://chromium.googlesource.com/infra/goma/server
```
## patch sources:
needed to disable oauth and enable compilation without docker or nsjail which I haven't tried yet
```
diff --git a/auth/acl/checker.go b/auth/acl/checker.go
index dc48b20..5bfdfc7 100644
--- a/auth/acl/checker.go
+++ b/auth/acl/checker.go
@@ -92,6 +92,8 @@ func (c *Checker) CheckToken(ctx context.Context, token *oauth2.Token, tokenInfo
 
 	logger := log.FromContext(ctx)
 
+    return "id1", nil, nil
+
 	for _, g := range c.config.GetGroups() {
 		if !checkGroup(ctx, tokenInfo, g, c.AuthDB) {
 			continue
@@ -132,6 +134,7 @@ func (c *Checker) CheckToken(ctx context.Context, token *oauth2.Token, tokenInfo
 func checkGroup(ctx context.Context, tokenInfo *auth.TokenInfo, g *pb.Group, authDB AuthDB) bool {
 	logger := log.FromContext(ctx)
 	logger.Debugf("checking group:%s", g.Id)
+    return true
 	if g.Audience != "" {
 		if tokenInfo.Audience != g.Audience {
 			logger.Debugf("audience mismatch: %s != %s", tokenInfo.Audience, g.Audience)
diff --git a/auth/client.go b/auth/client.go
index 3c12f1d..65572d3 100644
--- a/auth/client.go
+++ b/auth/client.go
@@ -124,6 +124,13 @@ func (a *Auth) Check(ctx context.Context, req *http.Request) (*enduser.EndUser,
 	defer span.End()
 	logger := log.FromContext(ctx)
 
+        fake_token := &oauth2.Token{
+                AccessToken: "",
+                TokenType:   "",
+        }
+        return enduser.New("fake_email", "fake_groupid", fake_token), nil
+
+
 	authorization := req.Header.Get("Authorization")
 	if authorization == "" {
 		logger.Warnf("no authorization header")
diff --git a/cmd/remoteexec_proxy/main.go b/cmd/remoteexec_proxy/main.go
index 97bd04a..11f9769 100644
--- a/cmd/remoteexec_proxy/main.go
+++ b/cmd/remoteexec_proxy/main.go
@@ -451,9 +451,6 @@ func main() {
 					RbeInstanceBasename: path.Base(*remoteInstanceName),
 					Properties: []*cmdpb.RemoteexecPlatform_Property{
 						{
-							Name:  "container-image",
-							Value: *platformContainerImage,
-						}, {
 							Name:  "OSFamily",
 							Value: "Linux",
 						},
diff --git a/command/descriptor/relocate.go b/command/descriptor/relocate.go
index 78a0b00..ec2cc04 100644
--- a/command/descriptor/relocate.go
+++ b/command/descriptor/relocate.go
@@ -93,7 +93,13 @@ func Relocate(setup *pb.CmdDescriptor_Setup, cmdpath string) (*pb.CmdDescriptor_
 	}
 	s := proto.Clone(setup).(*pb.CmdDescriptor_Setup)
 	origin := filepath.Dir(cmdpath)
-	s.CmdFile.Path = cmdpath
+//	s.CmdFile.Path = cmdpath
+	progpath, e := AbsCmdPath(s)
+	if e != nil {
+		return nil, e
+	}
+	s.CmdFile.Path = progpath
+
 	for _, f := range s.Files {
 		if filepath.IsAbs(f.Path) {
 			continue
diff --git a/remoteexec/exec.go b/remoteexec/exec.go
index f8bb143..9030f0f 100644
--- a/remoteexec/exec.go
+++ b/remoteexec/exec.go
@@ -552,6 +552,7 @@ func (r *request) newInputTree(ctx context.Context) *gomapb.ExecResp {
 	}
 
 	symAbsOk := r.f.capabilities.GetCacheCapabilities().GetSymlinkAbsolutePathStrategy() == rpb.SymlinkAbsolutePathStrategy_ALLOWED
+	symAbsOk = true
 
 	for _, f := range r.cmdFiles {
 		if _, found := toolchainInputs[f.Path]; found {
@@ -776,7 +777,7 @@ func (r *request) newWrapperScript(ctx context.Context, cmdConfig *cmdpb.Config,
 		wrapperData := digest.Bytes("wrapper-script", []byte(wrapperScript))
 		files, wrapperData = r.maybeApplyHardening(ctx, "InputRootAbsolutePath", files, wrapperData)
 		// https://cloud.google.com/remote-build-execution/docs/remote-execution-properties#container_properties
-		r.addPlatformProperty(ctx, "InputRootAbsolutePath", r.tree.RootDir())
+        //r.addPlatformProperty(ctx, "InputRootAbsolutePath", r.tree.RootDir())
 		for _, e := range r.gomaReq.Env {
 			envs = append(envs, e)
 		}
diff --git a/remoteexec/gcc.go b/remoteexec/gcc.go
index 10e34bd..7915df5 100644
--- a/remoteexec/gcc.go
+++ b/remoteexec/gcc.go
@@ -186,6 +186,9 @@ Loop:
 		case arg == "-idirafter":
 			pathFlag = true
 
+		case arg == "-fno-experimental-new-pass-manager":
+			continue
+
 		case strings.HasPrefix(arg, "-"): // unknown flag?
 			return unknownFlagError{arg: arg}
 
@@ -258,6 +261,8 @@ func clangArgRelocatable(filepath clientFilePath, args []string) error {
 			skipFlag = true
 		case arg == "-load":
 			pathFlag = true
+		case arg == "-fno-experimental-new-pass-manager":
+			continue
 		case strings.HasPrefix(arg, "-debug-info-kind="):
 		default:
 			return unknownFlagError{arg: fmt.Sprintf("clang: %s", arg)}
diff --git a/remoteexec/inputroot.go b/remoteexec/inputroot.go
index e6a37a0..a1f6c58 100644
--- a/remoteexec/inputroot.go
+++ b/remoteexec/inputroot.go
@@ -39,7 +39,9 @@ func commonDir(filepath clientFilePath, paths []string) string {
 	}
 	path := filepath.SplitElem(paths[0])
 	for _, p := range paths[1:] {
-		path = samePathElement(filepath, path, filepath.SplitElem(p))
+               if strings.HasPrefix(p, "/usr") == false {
+                       path = samePathElement(filepath, path, filepath.SplitElem(p))
+               }
 	}
 	return filepath.Join(path...)
 }
```
because of present of tabs and the copy paste, please remember to apply `--ignore-whitespace`
EDIT: updated for v90

## create a "config-file" file with:
```
version_id:"1",
configs: [{
        target: {
                addr:"localhost",
        },
        build_info: {
                creator:"uazo",
                hostname:"localhost",
                toolchain:"xx"
        },
        remoteexec_platform: {
                rbe_instance_basename: "default_instance",
                has_nsjail: false
        },
        dimensions: "os:linux"
}]
```

now, you can start goma server with:
```
export REDISHOST=localhost
/usr/local/go/bin/go run ~/server/cmd/remoteexec_proxy/main.go --port 5050 --remoteexec-addr localhost:50051 --remote-instance-name default_instance --insecure-remoteexec --allowed-users ppp --exec-config-file "config-file" --exec-check-cache-timeout 30s --exec-max-retry-count 50 --exec-execute-timeout 600s
```

the output should look like:
```
2021-03-29T10:03:49.202Z        INFO    remoteexec_proxy/main.go:295    allow access for ["ppp"] / domains []
2021-03-29T10:03:49.202Z        INFO    remoteexec_proxy/main.go:311    using default service account
2021-03-29T10:03:49.202Z        INFO    acl/checker.go:74       service account default: update
2021-03-29T10:03:49.202Z        INFO    acl/checker.go:83       acl updated
2021-03-29T10:03:49.219Z        WARN    remoteexec_proxy/main.go:395    use insecrure remoteexec API
2021-03-29T10:03:49.219Z        INFO    remoteexec_proxy/main.go:410    redis enabled for gomafile-digest: localhost:6379 idle=50 active=200
2021-03-29T10:03:49.220Z        INFO    exec/inventory.go:193   configure platform config: target:{addr:"localhost:50051"}  build_info:{creator:"uazo"  hostname:"localhost"  toolchain:"xx"}  remoteexec_platform:{rbe_instance_basename:"default_instance"}  dimensions:"os:linux"
```
# goma client installation:
```
cd ~
cipd install infra/goma/client/linux-amd64 -root ~/goma
```
## create the file for auth:
`echo "nomatter" >~/.debug_auth_file`

## modify ~/goma/goma_auth.py with:
```
#!/usr/bin/env python
def main():
  return 0
```
## set goma flags
```
export GOMA_SERVER_HOST=localhost
export GOMA_SERVER_PORT=5050
export GOMA_USE_SSL=false
export GOMA_HERMETIC=error
export GOMA_ARBITRARY_TOOLCHAIN_SUPPORT=true
export GOMA_HTTP_AUTHORIZATION_FILE=~/.debug_auth_file
export GOMA_USE_LOCAL=false
export GOMA_FALLBACK=true
```
you can test if goma is working with `~/goma/compiler_proxy`

# chromium build preparation:
```
gn gen --args="$(cat ~/bromite/build/GN_ARGS) target_cpu=\"arm64\" use_goma=true goma_dir=\"~/goma\" " out/arm64
```

# start compiling with:
```
~/goma/goma_ctl.py start
date && ninja -j16 -C out/arm64 chrome_public_apk && date
```
~~you can set how many local workers you can use with (8 in this example):~~

I have not found anything for setting the number of remote workers, goma use them as long as there are free.
unfortunately after 4, in my configuration, I get to saturate the connections, with `ConnectionError error: Concurrent RPC limit exceeded!`
if anyone knows how to fix I'm all ears!

**Comment by csagan5** [comment]
> unfortunately after 4, in my configuration, I get to saturate the connections, with `ConnectionError error: Concurrent RPC limit exceeded!`
> if anyone knows how to fix I'm all ears!

See https://stackoverflow.com/a/51609363; perhaps changing that `maximum_concurrent_rpcs` will fix this?

**Reply by uazo** [reply]
yes, we're almost there. in buildgrid the value corresponds to the `thread-pool-size` option, putting 50 the error disappears, and now i'm able to increase the number of bots.
I'm trying now with buildbox https://buildgrid.build/user/using_buildbox.html, which is the runner to use in production.

I just have to understand the reason for these times:
inventory: 0s, input tree: 1ms, setup: 0s, check cache: 0s, check missing: 2ms, upload blobs: 0s, **execute: 9.691s**, response: 55ms, total: 9.751s

**Reply by uazo** [reply]
I succeeded! just compiled bromite x86 in 02:34:34 using 20 cores on 4 machines (on lan)!

**Reply by csagan5** [reply]
Isn't that a bit too long (2h 34m) given the power at work?

**Reply by csagan5** [reply]
Regardless, congratulations!

Can you setup a repository and/or a guide on how to reproduce this?

**Reply by uazo** [reply]
> Isn't that a bit too long (2h 34m) given the power at work?

power I don't think so ... 4 servers with rather dated cpu.
However, consider that I start for the v90 from 7.30h with only 8 cores of those, halve using 16.

The compilation chain needs to be revised a little, because currently java is not covered by goma (although the code foresees it, is explicitly disabled), so at the beginning the first 15-20 minutes in which it compiles android are lost because goma is not used while on the contrary it could carry on with other projects.
but the way seems is that ...

> Can you setup a repository and/or a guide on how to reproduce this?

certainly, so can you help me on the setup of the tls and permissions. among other things, I think I need some sort of orchestrator, because I need to start/stop/restart processes on multiple servers and something that monitors them from a central place.

the steps are more or less those above, with v90 it must be modified in other points goma server because there are new compiler flags not managed in the published version, but not much stuff.
the most was figuring out how to compile buildbox (https://gitlab.com/BuildGrid/buildbox) and get it working, but in the end, once you understand, it's pretty straightforward

**Comment by uazo** [comment]
nothing to do, tried with two servers on the same lan, modifying thread-pool-size (150 is the right choice) and using buildbox, it works but it takes longer anyway, possibly due to ARBITRARY_TOOLCHAIN_SUPPORT (which also downloads toolchain executables) or perhaps because 16 bots are too few.
the goal of halving the time seems far away.

I begin to consider the idea that goma is more of a distributed cache than a compile speeder, that is, it speeds up if someone else has already compiled the same code... I have to check the "luci" logs to understand their current compile times.

at most I think it is probably possible to launch the build of at least v8 and blink (the heaviest and slowest) separately and in parallel from the rest of chromium, it seems a shame not to delve into the subject.

have you done / are you going to do some tests?

**Reply by csagan5** [reply]
> at most I think it is probably possible to launch the build of at least v8 and blink (the heaviest and slowest) separately and in parallel from the rest of chromium, it seems a shame not to delve into the subject.
> 
> have you done / are you going to do some tests?

I have over the years, and the best I could come up with are [mtool](https://github.com/bromite/mtool) (that you also have now) and using ccache; ccache [sometimes gives out weird linking errors (because of some bug in ccache)](https://github.com/bromite/bromite/issues/870) and requires cleaning the cache folders. Jumbo builds were helping but [they do not exist anymore in Chromium](https://github.com/chromium/chromium/commit/878b18040e2d5a6d2c39f5a49be6f9dc04c35d03).

v8 and blink are already building in parallel following how ninja works; since the Gandi servers are so slow (and without SSD) I am looking into a different solution, but no timeline yet.

In the end I think the best optimisation is to have 2 Docker images with pre-built Chromium and Bromite (sources + output folders); they will be huge but we build them (quickly or not, does not matter) on a separate (powerful) machine and then re-use them for PRs. All PRs will be able to use the pre-built material and even normal GitHub runners for the changes.

We could even try forking the chromium repo to have all the history and commits, but that is less important than fast PR builds.

Do you like the idea?

**Reply by uazo** [reply]
> In the end I think the best optimisation is to have 2 Docker images with pre-built Chromium and Bromite (sources + output folders); they will be huge but we build them (quickly or not, does not matter) on a separate (powerful) machine and then re-use them for PRs. All PRs will be able to use the pre-built material and even normal GitHub runners for the changes.
> Do you like the idea?

certainly it is an alternative, and I think it can be done regardless of the use of goma, which can be integrated after fully understanding the use, but I do not deny that my intention is to review all the patches to try to eliminate/reduce the "git am" mechanism that I hate, so I am trying to improve the performance of the builds, because I know I have to recompile everything and often before I can offer you a solution to evaluate.

> We could even try forking the chromium repo to have all the history and commits

that's not what I'm trying to do.
I would like to evaluate the possibility of having the bromite code externally to the chromium code, with its history, brave succeeds, why shouldn't I do it too? :)
I am convinced in bromite and I would like to be able to contribute to make it become better also from the point of view of development, as well as of use.

**Reply by csagan5** [reply]
> I would like to evaluate the possibility of having the bromite code externally to the chromium code, with its history, brave succeeds, why shouldn't I do it too? :)

I do not believe that the public repository we see from Brave is also what they use internally; I might be wrong though. I say this because of how the patches/commits look.

If we find a solution that allows people to make it easy to try out things the project will benefit from it.

**Reply by uazo** [reply]
> I do not believe that the public repository we see from Brave is also what they use internally; I might be wrong though. I say this because of how the patches/commits look.

well, I think so instead.. we have the mole, maybe he tells us ;)

> If we find a solution that allows people to make it easy to try out things the project will benefit from it.

of course!

**Reply by csagan5** [reply]
There is also https://wiki.debian.org/UsingQuilt and https://stacked-git.github.io/

**Reply by uazo** [reply]
> https://stacked-git.github.io/

very interesting. you use it?

**Reply by csagan5** [reply]
No, but I thought it was related and interesting.

**Comment by uazo** [comment]
excuse me if I have not made myself heard on the subject, now I have a bit clearer ideas

back then, it happened that to integrate the tests into the c.i. that I built, I had to change the machine, luckily even the new one as a gift.
this is because the processor did not allow me to start android.
now I use an 2xe5-2650 (32 core) with 16gb of ram: using that the benefit of having goma is 20-30 minutes less on the build! (not counting the cost of electricity). It's not worth it, as there are still things to fix.

so the question, as @csagan5 said, is power. probably goma would make sense if I could do a -j 1000 but I can't, not even using 3 servers at the same time (ah, and paralleling the build in many threads would allow that @csagan5, now i know for sure)

it's a great pity but to exploit goma well you need a lot of powerful servers, which I don't have.
so now I use the c.i. without goma.

for me it is a closed chapter, I have moved on to something else.
